Add Histogram split search supports for Oblique forests

ariellubonja opened the pull request:
Hi Richard, Mathieu!

This is minimal changes to add Histogramming support to Oblique forests

As for user-side toggling, we discussed w/ Richard whether decision_tree.proto:NumericalSplit should have an analogous version for only histograms. This requires further discussion

ariellubonja committed:
Add Histogram split search supports for Oblique forests

ariellubonja committed:
Add random engine to support Oblique Histogramming call

rstz commented on yggdrasil_decision_forests/learner/decision_tree/oblique.cc:
```diff
@@ -311,13 +312,25 @@ absl::StatusOr<SplitSearchResult> EvaluateProjection(
   // Find a good split in the current_projection.
   SplitSearchResult result;
   if constexpr (is_same<LabelStats, ClassificationLabelStats>::value) {
+    if (dt_config.numerical_split().type() == proto::NumericalSplit::EXACT) {
```
Can you please add fields to the oblique proto message https://github.com/google/yggdrasil-decision-forests/blob/main/yggdrasil_decision_forests/learner/decision_tree/decision_tree.proto#L165
for the type of split and check this field here (see my other comment)

ariellubonja replied:
Done, confirm below

Commit:
Plumb random as trailing EvaluateProjection parameter with default

EvaluateProjection's `random` lived in the middle of the parameter list
with no default, forcing every caller to pass it explicitly even on the
Cart path that doesn't need it. Move it to a trailing parameter with
`= nullptr` default in the header, and propagate the reorder through
the impl, the three explicit template instantiations,
EvaluateProjectionAndSetCondition, and the call sites in oblique.cc
and vector_sequence.cc. No behavior change.

Commit:
Dispatch oblique projection split via SparseObliqueSplit.projection_split

Per Richard's review: add a dedicated NumericalSplit field to
SparseObliqueSplit so the projection-axis split type is orthogonal to
the top-level DecisionTreeTrainingConfig.numerical_split (which only
controls axis-aligned numerical features). Setting
sparse_oblique_split.projection_split.type to HISTOGRAM_RANDOM or
HISTOGRAM_EQUAL_WIDTH routes the projected-value scan through the
histogram finder; absent or EXACT keeps the existing CART path.

Implementation notes:
- The histogram finder reads dt_config.numerical_split().{type,
  num_candidates}, so EvaluateProjection clones dt_config and
  overwrites its top-level numerical_split with the projection_split
  before calling. Avoids changing the histogram fn's signature, which
  would expand the diff into training.cc/h unnecessarily.
- MHLD callers (vector_sequence.cc) don't set sparse_oblique_split, so
  projection_split() returns a default NumericalSplit with type=EXACT;
  the MHLD path stays on the CART path with no behavior change.
- SetDefaultHyperParameters mirrors the existing num_candidates
  auto-default (1 for HISTOGRAM_RANDOM, 255 for HISTOGRAM_EQUAL_WIDTH)
  for the new projection_split field.
- Classification only in this commit; regression histogram dispatch is
  a mechanical follow-up (FindSplitLabelRegressionFeatureNumerical
  Histogram exists upstream).

ariellubonja commented:
Hi Richard! I would like to incorporate Random histogram SPORF into YDF going forward. I implemented the plumbing of the RNG generator `random`. Please take a look. Apologies for forgetting this PR open all this time

ariellubonja commented on yggdrasil_decision_forests/learner/decision_tree/training.cc:
```diff
@@ -4377,6 +4377,27 @@ void SetDefaultHyperParameters(proto::DecisionTreeTrainingConfig* config) {
+                   ->mutable_projection_split();
+    switch (ps->type()) {
+      case proto::NumericalSplit::HISTOGRAM_RANDOM:
+        ps->set_num_candidates(1);
```
I don't understand why in Axis-Aligned, this is set to 1 by default (training.cc:4370). 63 or 255 is more reasonable
```suggestion
        ps->set_num_candidates(63);
```

ariellubonja commented on yggdrasil_decision_forests/learner/decision_tree/oblique.cc:
```diff
@@ -312,23 +312,29 @@ absl::StatusOr<SplitSearchResult> EvaluateProjection(
+      // candidate thresholds from numerical_split. Apply values from
+      // projection_split
+      proto::DecisionTreeTrainingConfig histogram_dt_config = dt_config;
+      *histogram_dt_config.mutable_numerical_split() = projection_split;
```
Histogram needs to know n. bins, which it normally reads from `dt_config.numerical_split`. Thoughts? Copy cost is a small concern

ariellubonja replied:
[Alternative approach to skip copying config](https://github.com/ariellubonja/yggdrasil-oblique-forests/commit/fdbeff620dafc88f54aee0df5fec80ba2460a2fa)

ariellubonja replied:
Ariel: use alternate approach

ariellubonja replied:
Write 2-3 tests for Histogramming in `training_test.cc`: Toy dataset ~100 examples, single split - Sanity-check. 1 for Random, 1 for Equal Width. Follow examples in training_test.cc (see SparseObliqueClassification test).  [Opt] Integration test random_forest_test.cc - test on real data

ariellubonja commented:
I ran [Accuracy results](https://docs.google.com/spreadsheets/d/1GjEnI8p0CxJE_Q6CNe-Yypk06XYKkcFiKFeSxvz0IoY/edit?usp=sharing), and they informed my Suggestion to bump the default `num_candidates` from 1 to 64. Random Hist with 1 candidate (as default) has both a low accuracy on average and a high variance

<img width="624" height="547" alt="Screenshot 2026-08-30 at 15 35 17" src="https://github.com/user-attachments/assets/9381aef9-5065-4a3c-baaa-93293de52b18" />

ariellubonja commented on yggdrasil_decision_forests/learner/decision_tree/decision_tree.proto:
```diff
@@ -285,6 +285,10 @@ message DecisionTreeTrainingConfig {
     optional int32 max_num_features = 11 [default = -1];
+
+    // Controls how the split in the projection space is computed: EXACT (default)
+    // or HISTOGRAM_* (Random or Equal Width)
```
Update comment: Histogram Random splits are done on 64 bins. Equal Width is done on 256. These defaults might be different from Axis-Aligned splits

ariellubonja commented:
Hi Richard! As per our discussion earlier today, I added:

1. Avoided passing fake `dt_config` and instead pass `numerical_split` arg
2. Unit tests w/ a toy dataset of 100 examples
3. Integration tests - I used the same target accuracy as `RandomForestOnAdult.SparseOblique` (Exact) and `RandomForestOnAbalone.SparseOblique`
4. Extended Histograms to Regression. We didn't notice this part, but it applies identically. Added tests for it too

5. [Suggestion] I now suggest renaming `RandomForestOnAdult.SparseOblique` to `RandomForestOnAdult.SparseObliqueExact`. Thoughts?
